toolchain/check/import_ref.cpp

```diff
@@ -1988,10 +1982,43 @@ static auto TryResolveTypedInst(ImportRefResolver& resolver,
             class_const_inst.type_id());
     class_id = generic_class_type.class_id;
   }
+
+  // TODO: Ensure the vtable is only imported once, in eg: if there's distinct
+  // vtable constants (imported from multiple libraries using the vtable) that
+  // refer to the same vtable, the vtable should still be singular.
+  auto virtual_functions =
+      resolver.import_inst_blocks().Get(import_vtable.virtual_functions_id);
+
+  llvm::SmallVector<SemIR::InstId> lazy_virtual_functions;
+  lazy_virtual_functions.reserve(virtual_functions.size());
+  for (auto vtable_entry_id : virtual_functions) {
+    auto local_attached_constant_id = GetLocalConstantId(
+        resolver,
+        resolver.import_constant_values().GetAttached(vtable_entry_id));
+    if (resolver.HasNewWork()) {
+      continue;
+    }
+    // Use LoadedImportRef for imported symbolic constant vtable entries so they
+    // can carry attached constants necessary for applying specifics to these
+    // constants when they are used.
+    lazy_virtual_functions.push_back(
+        local_attached_constant_id.is_symbolic()
+            ? AddLoadedImportRef(
+                  resolver,
+                  GetSingletonType(resolver.local_context(),
+                                   SemIR::SpecificFunctionType::TypeInstId),
+                  vtable_entry_id, local_attached_constant_id)
+            : resolver.local_constant_values().GetInstId(
+                  local_attached_constant_id));
+  }
+  if (resolver.HasNewWork()) {
+    return ResolveResult::Retry();
+  }
```

Comment:
Rather than adding a second `HasNewWork` / `Retry()` here, we generally do all the work that could add new work before a single `HasNewWork` / `Retry()`. (There is a special case for declaration-like entities that need three-pass construction instead of two-pass, but it doesn't apply here.)

Applying that approach here, we'd make all the `GetLocalConstantId` calls before the `HasNewWork` / `Retry`, and then do all the `AddLoadedImportRef`s afterwards.

Reply:
Think I've addressed this in [ae68951](https://github.com/carbon-language/carbon-lang/pull/5871/commits/ae68951f90e32b0ea26f212a318a39303b6b6740)